### PR TITLE
[SRVKS-887] Add Knative services PersistentVolumeClaim ConfigMap

### PR DESCRIPTION
**Pull request by rh-max**
Version(s): 4.6+

Issue: https://issues.redhat.com/browse/SRVKS-887

Link to docs preview: https://deploy-preview-45083--osdocs.netlify.app/openshift-enterprise/latest/serverless/admin_guide/serverless-configuration.html#serverless-enabling-pvc-support_serverless-configuration

### CODE REVIEW

**Comment by rh-max** [comment]
@skonto Hi Stavros could you please have a look at this? Thanks.

**Review by jrangelramos — COMMENTED** [review]
I think the example is Ok, howerver I think it could be more interesting to end-users if the example claims for a volume which the service can actually write (and read) data. So that it should contain as well storage spec request such as 100mb, 500mb, 1gb, etc and the mount path.
wdyt?

**Comment by maschmid** [comment]
The docs fails to mention the feature flag that must be enabled, I'll see if I can find out how to configure them (there are `kubernetes.podspec-persistent-volume-claim` and `kubernetes.podspec-persistent-volume-write`  flags mentioned in upstream docs)

**Comment by rh-max** [comment]
@jrangelramos Can you please provide the YAML snippet for this scenario?
@maschmid Good point, I've added the info on the feature flags.

File: modules/serverless-enabling-pvc-support.adoc [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+...
+spec:
+ template:
+   spec:
+     containers:
+         ...
+         volumeMounts: <1>
+           - mountPath: /data
+             name: mydata
+             readOnly: true
```
Comment (skonto) [inline modules/serverless-enabling-pvc-support.adoc:44]: s/true/false

**Comment by skonto** [comment]
Also it is important to note that users might have to setup proper user rights when mounting a persistent volume with the Knative container user as with any use of a pv, but this is goes beyond this doc, and it is a generic K8s pv issue.
As one quick option though they could try setup the mount with a specific fsgroup set to the user with which we by default run the container `65532`. This is done via the pod security context fsGroup property.

**Comment by rh-max** [comment]
> Also it is important to note that users might have to setup proper user rights when mounting a persistent volume with the Knative container user as with any use of a pv, but this is goes beyond this doc, and it is a generic K8s pv issue. As one quick option though they could try setup the mount with a specific fsgroup set to the user with which we by default run the container `65532`. This is done via the pod security context fsGroup property.

Thanks. I've mentioned the user permissions aspect, but there's much more to persistent storage, so I've added a link to persistent storage docs.

> I think the example is Ok, howerver I think it could be more interesting to end-users if the example claims for a volume which the service can actually write (and read) data. So that it should contain as well storage spec request such as 100mb, 500mb, 1gb, etc and the mount path. wdyt?

Can you @jrangelramos @maschmid or @skonto please provide the changes needed for this Jefferson's suggestion? Thank you.

**Comment by skonto** [comment]
> I think the example is Ok, howerver I think it could be more interesting to end-users if the example claims for a volume which the service can actually write (and read) data. So that it should contain as well storage spec request such as 100mb, 500mb, 1gb, etc and the mount path. wdyt?

Using the default storage class users can specify a pvc request as follows:

```
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: knative-pv-claim
  namespace: serving-tests
spec:
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 1Gi
```

**Comment by maschmid** [comment]
I don't think a ReadWriteOnce would be a suitable mode for a knative service (as that is by default scalable)?

**Comment by skonto** [comment]
I was still editing this. :)

**Comment by skonto** [comment]
Hast his been tested enough on OCP? Using ceph or some other setup?

**Comment by rh-max** [comment]
@skonto Thanks a bunch. I've added this config, could you please give it another glance?

> @maschmid Has this been tested enough on OCP? Using ceph or some other setup?

If all is fine, and if it has been tested enough (or lack of testing doesn't preclude these docs), I'll send it on for docs team review.

**Comment by maschmid** [comment]
> @maschmid Has this been tested enough on OCP? Using ceph or some other setup?
yes, I have verified this manually with ceph (using OpenShift Data Foundation operator on AWS, with `ocs-storagecluster-cephfs` StorageClass)

**Comment by skonto** [comment]
Awesome thanks @maschmid

File: modules/serverless-enabling-pvc-support.adoc [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    requests:
+      storage: 1Gi
+----
+
+.Knative service PVC configuration
+[source,yaml]
+----
+apiVersion: serving.knative.dev/v1
+kind: Service
+...
```
Comment (skonto) [inline modules/serverless-enabling-pvc-support.adoc:51]: Add the namespace here to make clear that pvc and service need to have the same value?

File: serverless/admin_guide/serverless-configuration.adoc [hunk 3]
```diff
@@ -24,6 +24,15 @@ include::modules/serverless-https-redirect-global.adoc[leveloffset=+1]
...
 // Kourier Gateway service type
 include::modules/serverless-kourier-gateway-service-type.adoc[leveloffset=+1]
+// Enabling PVC for Serving
+include::modules/serverless-enabling-pvc-support.adoc[leveloffset=+1]
+
+ifdef::openshift-enterprise[]
+[id="additional-resources_serverless-enabling-pvc-support"]
+[role="_additional-resources"]
+== Additional resources
+* xref:../../storage/understanding-persistent-storage.html#understanding-persistent-storage[Understanding persistent storage]
```
Comment (bergerhoffer) [inline serverless/admin_guide/serverless-configuration.adoc:34]: .html -> .adoc

File: modules/serverless-enabling-pvc-support.adoc [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+//
+// * serverless/admin_guide/serverless-configuration.adoc
+
+:_content-type: REFERENCE
+[id="serverless-enabling-pvc-support_{context}"]
+= Enabling PVC support
+
+Some serverless applications need permanent data storage. To achieve this, you can configure Persistent Volume Claims (PVCs) for your Knative services.
+
+* The `"kubernetes.podspec-persistent-volume-claim"` extension controls whether Persistent Volumes (PVs) can be used with Knative Serving.
```
Comment (bergerhoffer) [inline modules/serverless-enabling-pvc-support.adoc:11]: Persistent Volumes -> persistent volumes

**Review by bergerhoffer — COMMENTED** [review]
A few comments!

File: modules/serverless-enabling-pvc-support.adoc [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+.PersistentVolumeClaim configuration
+[source,yaml]
+----
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: example-pv-claim
+  namespace: my-ns
+spec:
```
Comment (skonto) [inline modules/serverless-enabling-pvc-support.adoc:38]: Users should use a storage class here as well that supports the specific access mode. For example if users run on gcloud and use the default one which is `gp2` the pvc will never be ready as it does not support ReadWriteMany.
So let's add that as a note and the field: `storageClassName`:  ...`.

File: modules/serverless-enabling-pvc-support.adoc [hunk 6]
```diff
@@ -0,0 +1,74 @@
+// Module included in the following assemblies:
+//
+// * serverless/admin_guide/serverless-configuration.adoc
+
+:_content-type: REFERENCE
+[id="serverless-enabling-pvc-support_{context}"]
+= Enabling PVC support
```
[the remark this answers is not in the document]
Comment (rh-max) [reply modules/serverless-enabling-pvc-support.adoc:7]: I generally agree with this. However, the whole Global configuration section is currently in the format of short reference sections, each giving general snippets for some piece of available functionality - as opposed to proper procedures.
AFAIK, we intend to rewrite the whole section as procedures, so "Enabling PVC support" be addressed too.

**Comment by jrangelramos** [comment]
LGTM from QE side

**Pull request merged**